wpewebkit is not compiling in latest master for Westeros (upstream problem)

Hi guys,
I went back to my habit of building `meta-wpe` regularly and I noticed that the `wpe-westeros-image` is not building anymore, failing to compile `wpewebkit`. It looks like it's an upstream problem.
I'm raising this issue in case you haven't noticed yet.  The details below:

```
Build Configuration:
BB_VERSION           = "1.37.0"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "universal"
TARGET_SYS           = "arm-poky-linux-gnueabi"
MACHINE              = "raspberrypi0-wifi"
DISTRO               = "poky"
DISTRO_VERSION       = "2.5"
TUNE_FEATURES        = "arm armv6 vfp arm1176jzfs callconvention-hard"
TARGET_FPU           = "hard"
meta
meta-poky
meta-yocto-bsp       = "master:d886fa118c930d0e551f2a0ed02b35d08617f746"
meta-oe
meta-python
meta-multimedia
meta-networking      = "master:d59b9806f743cea0168cddf942a31dcf446839e6"
meta-raspberrypi     = "master:de597abc3b9447611951469ad1ad18d5068ff074"
meta-wpe             = "master:f9c5f17c5acf835fd3166f5b6ca069a4b1e01ac5"
```

The `local.conf` file:

```
MACHINE = "raspberrypi0-wifi"
SERIAL_CONSOLE = ""
DISABLE_RPI_BOOT_LOGO = "1"
DISABLE_OVERSCAN = "1"
DISABLE_SPLASH = "1"
BOOT_DELAY = "0"
GPU_MEM_256 = "128"
GPU_MEM_512 = "196"
GPU_MEM_1024 = "396"
DISTRO_FEATURES_remove = "x11"
DISTRO_FEATURES_append = " opengl"
DISTRO_FEATURES_append = " systemd"
VIRTUAL-RUNTIME_init_manager = "systemd"
VIRTUAL-RUNTIME_initscripts = ""
DISTRO_FEATURES_BACKFILL_CONSIDERED += "sysvinit"
PACKAGE_CLASSES = "package_ipk"
IMAGE_INSTALL_append = " linux-firmware-bcm43430 wpa-supplicant"
IMAGE_FSTYPES = "tar.bz2 ext4 rpi-sdimg"
SDIMG_ROOTFS_TYPE = "ext4"
```

And the compilation warnings/error:

```
/home/yocto/wpe/tmp/work/arm1176jzfshf-vfp-poky-linux-gnueabi/wpewebkit/2.20+gitAUTOINC+7059b7c7b0-r1/git/Source/bmalloc/bmalloc/Gigacage.h:136:59: warning: cast from 'char*'
 to 'Gigacage::BasePtrs*' increases required alignment of target type [-Wcast-align]
     return *reinterpret_cast<BasePtrs*>(g_gigacageBasePtrs);
                                                           ^
{standard input}: Assembler messages:
{standard input}:328: Error: invalid literal constant: pool needs to be closer
{standard input}:329: Error: invalid literal constant: pool needs to be closer
{standard input}:334: Error: invalid literal constant: pool needs to be closer
{standard input}:335: Error: invalid literal constant: pool needs to be closer
{standard input}:340: Error: invalid literal constant: pool needs to be closer
{standard input}:341: Error: invalid literal constant: pool needs to be closer
{standard input}:346: Error: invalid literal constant: pool needs to be closer
{standard input}:347: Error: invalid literal constant: pool needs to be closer
{standard input}:352: Error: invalid literal constant: pool needs to be closer
{standard input}:353: Error: invalid literal constant: pool needs to be closer
{standard input}:358: Error: invalid literal constant: pool needs to be closer
{standard input}:359: Error: invalid literal constant: pool needs to be closer
{standard input}:364: Error: invalid literal constant: pool needs to be closer
{standard input}:365: Error: invalid literal constant: pool needs to be closer
{standard input}:371: Error: invalid literal constant: pool needs to be closer
{standard input}:377: Error: invalid literal constant: pool needs to be closer
```

There are a lot of these same casting warnings, but this one is the only one with an actual error from the assembler compiler.

finally we are having some good news by disabling the JIT for armv6 :)
now wpewebkit compiles! ..but doesn't install :( The error is:
```
UnifiedSource69.cpp:(.text._ZN3JSC25uniqueSecurityOriginTokenEv+0x20): undefined reference to `__atomic_fetch_add_8'
collect2: error: ld returned 1 exit status
```
I found a similar issue here https://github.com/PurpleI2P/i2pd/issues/897 that suggests adding `-latomic` to the linker. What do you think? Probably we'll need to add libatomic as build dependency. I will try to test this somehow. Open to your suggestions anyway @kraj !

@wouterlucas too sad to read that, my main motivation to work with meta-wpe is that it was the only browser that worked decently in the RPI0 for a personal project. I understand the RPI0/1 is not a powerful machine but supporting it was a very big plus for me :(  I will open an issue in WPEWebKit depending on if we give-up here, hopefully they care to keep it working :)

--- a/recipes-wpe/wpewebkit/files/0003-Improve-checking-if-libatomic-is-needed-for-armv6-pr.patch
+++ b/recipes-wpe/wpewebkit/files/0003-Improve-checking-if-libatomic-is-needed-for-armv6-pr.patch
@@ -1,0 +1,53 @@
+From 5ddafade7ecbb47d751c9d2af33c7b68832695a9 Mon Sep 17 00:00:00 2001
+From: Hugo Hromic <hhromic@gmail.com>
+Date: Fri, 6 Jul 2018 23:11:54 +0100
+Subject: [PATCH] Improve checking if libatomic is needed for armv6 processors.
+
+From Khem Raj:
+> This patch is a correct way of detecting feature support for atomics, that does not
+> depend upon architecture but on compiler being able to support it.
+> Different compilers can provide different kind of support e.g. these atomics may not
+> exist on i586 for clang but might for gcc and so on.
+
+Fixes:
+
+UnifiedSource69.cpp:(.text._ZN3JSC25uniqueSecurityOriginTokenEv+0x20): undefined reference to `__atomic_fetch_add_8'
+collect2: error: ld returned 1 exit status
+
+during do_install() for armv6, e.g. raspberrypi0 machine.
+
+Upstream-Status: Pending
+Signed-off-by: Hugo Hromic <hhromic@gmail.com>
+---
+ Source/JavaScriptCore/CMakeLists.txt | 17 +++++++++--------
+ 1 file changed, 9 insertions(+), 8 deletions(-)
+
+diff --git a/Source/JavaScriptCore/CMakeLists.txt b/Source/JavaScriptCore/CMakeLists.txt
+index 26e3b5388d7..9a7948393b2 100644
+--- a/Source/JavaScriptCore/CMakeLists.txt
++++ b/Source/JavaScriptCore/CMakeLists.txt
+@@ -116,15 +116,16 @@ set(JavaScriptCore_LIBRARIES
+     ${LLVM_LIBRARIES}
+ )
+ 
+-# Since r228149, on MIPS we need to link with -latomic, because
+-# __atomic_fetch_add_8 is not available as a compiler intrinsic. It is
+-# available on other platforms (including 32-bit Arm), so the link with
+-# libatomic is only neede on MIPS.
+-if (WTF_CPU_MIPS)
+-    list(APPEND JavaScriptCore_LIBRARIES
+-        -latomic
+-    )
++# Check if libatomic is needed in order to use std::atomic, and add
++# it to the list of JavaScriptCore libraries.
++file(WRITE ${CMAKE_BINARY_DIR}/test_atomic.cpp
++    "#include <atomic>\n"
++    "int main() { std::atomic<int64_t> i(0); i++; return 0; }\n")
++try_compile(ATOMIC_BUILD_SUCCEEDED ${CMAKE_BINARY_DIR} ${CMAKE_BINARY_DIR}/test_atomic.cpp)
++if (NOT ATOMIC_BUILD_SUCCEEDED)
++    list(APPEND JavaScriptCore_LIBRARIES -latomic)
+ endif ()
++file(REMOVE ${CMAKE_BINARY_DIR}/test_atomic.cpp)
+ 
+ set(JavaScriptCore_SCRIPTS_SOURCES_DIR "${JAVASCRIPTCORE_DIR}/Scripts")
+ 

--- a/recipes-wpe/wpewebkit/wpewebkit_2.20.bb
+++ b/recipes-wpe/wpewebkit/wpewebkit_2.20.bb
@@ -102,6 +102,9 @@ ARM_INSTRUCTION_SET_armv7r = "thumb"
 ARM_INSTRUCTION_SET_armv7m = "thumb"
 ARM_INSTRUCTION_SET_armv7ve = "thumb"
 
+# JSC JIT doesn't currently compile on ARMv6, disable it.
+EXTRA_OECMAKE_append_armv6 = " -DENABLE_JIT=OFF"
+
 do_compile() {
     ${STAGING_BINDIR_NATIVE}/ninja ${PARALLEL_MAKE} -C ${B} libWPEWebKit.so libWPEWebInspectorResources.so WPEWebProcess WPENetworkProcess WPEStorageProcess WPEWebDriver
 }

--- a/recipes-wpe/wpewebkit/wpewebkit_2.20.bb
+++ b/recipes-wpe/wpewebkit/wpewebkit_2.20.bb
@@ -25,7 +25,8 @@ SRC_URI = "${BASE_URI} \
            file://0001-Define-MESA_EGL_NO_X11_HEADERS-when-not-using-GLX.patch \
            file://0001-Fix-build-with-musl.patch \
            file://0002-include-GraphicsContext3D.h-for-DONT_CARE-definition.patch \
-          "
+           file://0003-Improve-checking-if-libatomic-is-needed-for-armv6-pr.patch \
+           "
 
 S = "${WORKDIR}/git"
 

